feat(builtins): implement regex (#138)
open-policy-agent/swift-opa: CHANGELOG.md

## Unreleased

> ### `regex` builtins (RE2-only contract)

Reviewer: ```suggestion
### `regex` builtins
```
[nit]: Let's just call out that the builtins are here, and then explain the "RE2 is the only supported syntax" bits in the description below".

Author: Fixed

Swift OPA now supports the `regex.*` family of builtins:

- `regex.is_valid`
- `regex.match`
- `regex.replace`
- `regex.split`
- `regex.find_n`
- `regex.find_all_string_submatch_n`
- `regex.template_match`

> > **Heads up — RE2 is the long-term contract.** Only Go [RE2](https://github.com/google/re2/wiki/Syntax) syntax is supported, and only RE2 syntax is *guaranteed* to keep working in future releases. Patterns that happen to work today purely because the current implementation is built on `NSRegularExpression` (ICU) — for example backreferences, lookaround, atomic groups, or ICU-style named groups — are not part of the contract and are not guaranteed to keep working. **Do not rely on non-RE2 features.**

Reviewer: ```suggestion
> [!WARNING]
> Swift OPA's `regex.*` builtins are only planned to support [RE2](https://github.com/google/re2/wiki/Syntax) syntax in the long term. Any regex syntax outside of RE2 syntax is unsupported, and may be broken in a future release.
```
[suggestion]: We can copy the warning block over from the README wholesale, and Github *should* render the `WARNING` block correctly in the release notes later on. 🤔

Author: Fixed

> To enforce this, every pattern is validated against an RE2 compatibility scanner up-front. Patterns using backreferences (`\1`–`\9`), lookahead (`(?=...)`, `(?!...)`), lookbehind (`(?<=...)`, `(?<!...)`), or atomic groups (`(?>...)`) are rejected at compile time — `regex.is_valid` returns `false`, and the other builtins throw. The scanner will be tightened over time. See the README's "Regex syntax — RE2 only" section for the full contract.

Reviewer: ```suggestion
To enforce this, every pattern is validated against an RE2 compatibility scanner up-front. Patterns using backreferences (`\1`–`\9`), lookahead (`(?=...)`, `(?!...)`), lookbehind (`(?<=...)`, `(?<!...)`), or atomic groups (`(?>...)`) are rejected at compile time: `regex.is_valid` returns `false`, and the other builtins throw. The scanner will be tightened over time. See the README's "Regex syntax support" section for more details.
```
[nit]: Small tweaks + corrected section ref.

Author: Fixed, thank you!
